### PR TITLE
Add Budget for EUR in Projects

### LINKED ISSUE #206: [BUG] Budget filters work with different currencies
On the main dashboard, the budet filter works with the "budget_org" value for initiatives. This is fine for "your" initiatives, as they will all be in the default currency for the selected organisation. But other organisations' initiatives will have their own default currency. So right now we are comparing different currencies... 

The ideal solution is to include the currency for each initiative in *all* of the default org currencies on the platform. This is quite a challenge, but I'm not sure of another way to be able to effectively group initiatives by 'budget' when every org has potentially a different currency.

### PR DESCRIPTION
**Pull request by dan-tang-ssd**
This PR is submitted to fix #206 

This PR contains below changes:
1. Add column projects.budget_eur
2. Add a command to calculate budget_eur if it is zero
3. Update dashboard stored procedure, use column projects.budget_eur instead of projects.budget in SQL for filtering
4. Projects create / edit view, calculate budget_eur automatically
5. Project import program, calculate budget_eur automatically

### CODE REVIEW

File: app/Http/Controllers/Admin/ProjectCrudController.php [hunk 1]
```diff
@@ -417,6 +426,26 @@ public function reAssess($id)
...
+        $this->calculateBudgetEur();
+
+        return $this->traitUpdate();
+    }
+
+    public function calculateBudgetEur() {
+        $budget = $this->crud->getRequest()->budget;
+        $exchangeRate = $this->crud->getRequest()->exchange_rate;
+        $this->crud->getRequest()->request->set('budget_eur', $budget * $exchangeRate);
+    }
```
Comment (dave-mills) [inline app/Http/Controllers/Admin/ProjectCrudController.php:447]: The exchange rate given  in the request will be from the project's currency to the organisation's primary currency - which is not necessarily EUR.

File: app/Imports/ProjectImport.php [hunk 2]
```diff
@@ -82,6 +82,7 @@ public function onRow(Row $row)
             'currency' => $data['currency'],
             'exchange_rate' => $data['exchange_rate'],
             'budget' => $data['budget'],
+            'budget_eur' => $data['budget'] * $data['exchange_rate'],
```
Comment (dave-mills) [inline app/Imports/ProjectImport.php:85]: Again, this is assuming the organisation's main currency is EUR.

**Review by dave-mills — CHANGES_REQUESTED** [review]
I don't think the `budget_eur` calculation is correct. the `exchange_rate` for a project is the exchange rate from the `$project->currency` to the `$organisation->currency`, not from the `$project->currency` to EUR. So we need a new exchange rate for each project.

**Comment by dan-tang-ssd** [comment]
Thank you Dave for your review and comments.

I will add new column projects.exchange_rate_eur.

By the way, I have a feeling that it will be a bit complicated, when we need to get exchange rate from a third party...
I have cloned live db to my local db, I did a checking on exchange_rates table with below SQL:

select distinct date from exchange_rates where date > '2023-01-01' order by date;

We only have exchange rate data for 189 / 365 days. 

Besides, sometimes the third party is not able to provide all 1089 (33 * 33 currencies) exchange rate data.
There are 1 and 759 records for 2023-06-30 and 2023-08-25 respectively.

Currently we have a daily schedule job run at 6AM to get yesterday exchange rate.
I checked that the queue is setup in live env in Forge, not sure why there is no exchange rate data after 2023-09-03...

---

I think it is better not to assume we always have exchange rate data available in applications.
A real example is user may create a new project with future start date.
So, I would suggest to allow user to enter exchange_rate_eur on both front end and projects import file. (make them as compulsory fields)
It would be the easiest way to obtain exchange_rate_eur, then calculate budget_eur so this project is available for filtering in dashboard. (only if user specify budget criteria)

For existing projects without exchange_rate_eur and budget_eur, we may need two new schedule jobs.
1. Find projects without exchange_rate_eur and budget_eur, get unique project start dates, retrieve exchange rate data for each project start date.
2. Find projects without exchange_rate_eur and budget_eur, find exchange rate for project currency to EUR, calculate budget_eur.

---

Proposed changes:
1. Add column projects.exchange_rate_eur
2. Initiative create / edit view, add compulsory field exchange_rate_eur for user to fill in. Hide it and set it to 1.0 if project currency is EUR.
3. Initiative import excel file template, add compulsory field exchange_rate_eur for user to fill in.
4. Add a command to re-retrive exchange rate data for project start date, if exchange rate data is missing or incomplete
5. Add a command to handle initiatives without exchange_rate_eur and budget_eur, find exchange rate data in local db for processing.

**Comment by dave-mills** [comment]
> Proposed changes:
> - Add column projects.exchange_rate_eur
> - Initiative create / edit view, add compulsory field exchange_rate_eur for user to fill in. Hide it and set it to 1.0 if project currency is EUR.

- I think you're right - it will be useful to show that we need this information to users and ask for it. I think we should approach it the same as the existing exchange rate - where the user can click the button to get the stored exchange rate if it exists, and also have the option of editing it. 
- I'm ultra-cautious about adding new compulsory fields, but I think you're right - we need to have this in order to properly filter the project on the dashboard. If the project currency is EUR you're right - we can hide it and set it to 1.0. Also, if the org's currency is EUR, we can hide it and set it to the main projects.exchange_rate value. 

> - Initiative import excel file template, add compulsory field exchange_rate_eur for user to fill in.
> - Add a command to re-retrive exchange rate data for project start date, if exchange rate data is missing or incomplete
> - Add a command to handle initiatives without exchange_rate_eur and budget_eur, find exchange rate data in local db for processing.

Agreed. Also, we should work out why the current job isn't working.

**Comment by dan-tang-ssd** [comment]
> Agreed. Also, we should work out why the current job isn't working.

I remote login to server, I can manually run the command to retrieve yesterday exchange rate successfully. (Only 330 / 1089 records retrieved).

I suspect there is something wrong with the queue. I have restarted the queue in Forge. I will check again next Monday. Let's see whether we have exchange rate data retrieved over this weekend.

Um... I am thinking whether we should have a monitoring mechanism to inform us when it detects a schedule job is not working.

**Comment by dave-mills** [comment]
(Only 330 / 1089 records retrieved).

I see 1089 records with 'yesterday'. Remember that it's 1 job per 33 currencies, and it rate-limits itself to avoid hitting the API limits, so it takes some time to get all of them. 

I agree re: monitoring. I also want to re-start our investigation into broader monitoring (e.g. we were looking at Laravel Flare a while ago. I want to restart that train of work and get something going). Lets talk later.

**Comment by dan-tang-ssd** [comment]
I added new column exchange_rate_eur in Initiative Import Excel Template File.

Here is the excel template file for testing:
[Agroecology Funding Tool - Initiative Import Template.xlsx](https://github.com/stats4sd/aec_portfolio/files/14221147/Agroecology.Funding.Tool.-.Initiative.Import.Template.xlsx)

I have also submitted PR #251 to update this file to dev branch.

**Comment by dan-tang-ssd** [comment]
I have finished my proposed changes. I think it is now ready for review.

**Review by dave-mills — APPROVED** [review]

**Review by dave-mills — APPROVED** [review]
Looks good. Seems to show and hide the relevant exchange rate fields and button at the right times based on org and project currency. 

I made a few tweaks to the wording, to add the reason for the EUR exchange rate and to clarify which rate is being asked for in each entry box. 

I think this is good to go.

**Pull request merged**